### ISSUE TITLE
TypeScript linting works only partially

### ISSUE THREAD

**Issue by ArmorDarks** [standard/standard#1283]
**What version of standard?**

v12.0.1

**What operating system, Node.js, and npm version?**

Window 10
Node v10.15.3
NPM v6.9.0

**What did you expect to happen?**

Expected to have ability use TypeScript linting

**What actually happened?**

Following #1278, provided docs are insufficient to make TypeScript linting fully work, since while it will become enabled, due to lack of the TypeScript plugin rules extensions, rules actually won't work (see https://github.com/standard/standard/pull/1278#issuecomment-495760576).

More of it, since there is no way to disable `no-unused-vars`, Standard will always fail on imports of the types.

For now there is no way to make it fully work. False-unsued vars can be fixed by providing within each file

```ts
/* eslint no-unused-vars: "off" */
/* eslint @typescript-eslint/no-unused-vars: "error" */
```

but that doesn't looks like convinient at all.

**Comment by shavyg2** [comment]
just ran into this too

**Comment by eladchen** [comment]
Just run into this as well.

**Comment by despian** [comment]
+1 also having problems with this

Another issue I came across is the indentation rules will break under certain circumstances.

You can fix this by adding:
```js
/* eslint indent: off */
/* eslint @typescript-eslint/indent: [error, 2] */
```

I'm sure there are probably also many other weird edge-case issues.

[typescript-eslint](https://github.com/typescript-eslint/typescript-eslint) does provide a set of sane rules [recommended.json](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/configs/recommended.json) but it's not possible to enable these when using standard.

Also, some of these rules aren't really compatible with the standard rules. E.g.
```js
    "@typescript-eslint/indent": "error"  // defaults to 4 space
```

It would be nice if typescript had proper "no config needed" support in standard.

Maybe the standard project could include their own set of sane defaults for typescript which are compatible with the normal js rules and apply them if the typescript plugin is loaded.

If I were to submit a PR along these lines is it likely it would be accepted?

**Comment by despian** [comment]
Actually, #1007 would resolve this.

**Comment by bjufre** [comment]
Any news here? This would make life much easier...

**Comment by PandawanFr** [comment]
Any updates?

I thought I could just add this to my standard config but I get `no-unused-vars`.
```json
"parser": "@typescript-eslint/parser",
"plugins": [
  "@typescript-eslint/eslint-plugin"
]
```

I know we can use `eslint-config-standard-with-typescript` but I think it'd be nice if it was supported by default in standard. I guess we have to wait for #1007 and/or #1321

**Comment by feross** [comment]
I'd like to have solve this in a future version of `standard`. In the meantime, you can use [`standardx`](https://www.npmjs.com/package/standardx):

1. `npm install standardx`
2. Rename the `"standard"` field in your `package.json` to `"standardx"`
3. Add a new `"rules": []` section with the following rules:

```json
"rules": {
  "no-unused-vars": "off",
  "@typescript-eslint/no-unused-vars": "error"
}
```

**Comment by ArmorDarks** [comment]
@feross Thanks for the reply

That was plan B, but it wasn't possible because `standardx` didn't work with VS Code extension. Now since you've merged https://github.com/standard/vscode-standardjs/pull/71 that should become possible after release.

**Comment by ndom91** [comment]
@feross would it make sense to open an issue on Typescript support in general and maybe even pin it? People can weigh in on what the best path forward would be (out of the box support, plugin, etc.), which rules to include, etc. etc.

**Comment by ThisIsManta** [comment]
Refer to [_Can I use a JavaScript language variant, like Flow or TypeScript?_](https://standardjs.com/index.html#typescript), I don't see `"extends": ["plugin:@typescript-eslint/recommended"]`. How do the rules offered by `@typescript-eslint/eslint-plugin` work? As far as I know, Standard does not allow us to specify `rules` or `extends` field, right?

I'm asking the above questions because I believe the problem here will be fixed if StandardJS allows me to `extends` or enable the rule [`@typescript-eslint/no-unused-vars`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/no-unused-vars.ts) (Do not be confused with the original rule [`no-unused-vars`](https://eslint.org/docs/rules/no-unused-vars) from ESLint)

**Comment by ahimta** [comment]
I think this issue is very important because It affects almost all Typescript users and requires configuration changes :sweat_smile:.

I understand that `standard` is mostly focused on pure Javascript and its ESLint & `standardx` siblings are intended to be more flexible and handle some edge-cases like this.

Let's assume only a dozen of developers are affected by this and spent 15 minutes (focused on finding a solution and confused by the behavior) each to mitigate it. Then the total time would be `12 * 15 = 180minutes = 3hours`.

In addition, the maintainers are spending focused time reading, understand, and replying to this.

IMHO, a maintainer should save everyones' time and first imagine that this is the only issue in the issue-tracker, allocate 10 minutes to think of fixes/mitigations, 10 minutes to choose one of them, and 30+ minutes to implement the fix/mitigation.

TLDR; :point_down: 

From the 30 seconds I skimmed this issue I think there are two fixes/mitigations:sweat_smile::
1. Document @feross `standardx` mitigation/workaround in FAQ (the FAQ item maybe added as the last comment before closing the issue).
2. Implement an actual fix for Typescript.

**Comment by toddbluhm** [comment]
Just wanted to put this package out there [`ts-standard`](https://github.com/toddbluhm/ts-standard) for anyone wanting to try it out. Its relatively new and I am not sure if this resolves all the issues perfectly, but it uses the `eslint-config-standard-with-typescript` rules and the `standard-engine` package under the hood with minimal required configuration. This is just another option to native typescript support in standard.

**Comment by ahimta** [comment]
Just made a [PR](https://github.com/standard/standard/pull/1434) with a working Typescript setup in the readme and hope it can close this issue:sweat_smile:.

**Comment by ArmorDarks** [comment]
Thanks! But, well, if we plan to go `standardx`-way, there are few issues at stake:

1. https://github.com/standard/vscode-standardjs/pull/71 has been merged, but not released. That means VS Code users won't be able to use it at all. I believe other popular editors might be affected by the same issue.
2. Even if `standard.engine` option will be released, it still requires additional configuration. That is, another developer will open the project with `standardx` and won't have any clue why VS Code doesn't show any errors. Proposal: https://github.com/standard/vscode-standardjs/issues/93

**Comment by SherloxFR** [comment]
For Typescript I find it better to use TSLint with the [tslint-config-standard](https://github.com/blakeembrey/tslint-config-standard) plugin.

It's really easy to setup:
```bash
yarn add tslint tslint-config-standard --dev
```
Create a `tslint.json` file with the following content:
```json
{
  "extends": [
    "tslint:recommended",
    "tslint-config-standard"
  ]
}
```
Then you can just call `tslint` from your CLI / `package.json` scripts to validate the syntax.

For the VSCode support be careful to use `TSLint (deprecated)` by `egamma`, because the one by Microsoft doesn't support autofix :sweat_smile:

**Comment by PandawanFr** [comment]
Although much easier, I don't think it's the best idea to promote TSLint when [it has been deprecated](https://github.com/palantir/tslint#tslint). The TSLint team are currently porting all of the TypeScript rules and logic over to [typescript-eslint](https://github.com/typescript-eslint/typescript-eslint) to make one unified linter for both JS and TS.

**Comment by david-szabo97** [comment]
Any update?

**Comment by SherloxFR** [comment]
@david-szabo97 as @PandawanFr said, [typescript-eslint](https://github.com/typescript-eslint/typescript-eslint) is the way to go. It works very well and allows to use any ESLint rule in addition to the Typescript specific rules.
Hit me on my [Discord](https://discord.gg/yQsKbwx) if you need any help to set it up.

**Comment by mightyiam** [comment]
Standard has rules for TypeScript, as well. And that repository has instructions on set up, as well:
https://github.com/standard/eslint-config-standard-with-typescript.

**Comment by JeffSpies** [comment]
Hi, all; what is the latest recommendation re TypeScript?

1) https://github.com/standard/eslint-config-standard-with-typescript as @mightyiam recommends above
2) StandardX as recommended on the [README](https://github.com/standard/standard/blame/master/README.md#L543) by @ahimta 

Thanks!

**Comment by LinusU** [comment]
@JeffSpies my personal recommendation is [`ts-standard`](https://github.com/toddbluhm/ts-standard), which will probably be [transfered to the Standard org.](https://github.com/toddbluhm/ts-standard/issues/14) in a not so distant future ☺️

**Comment by Divlo** [comment]
`ts-standard` is awesome! We should update the [README.md](https://github.com/standard/standard#user-content-typescript), that everyone using TypeScript should use `ts-standard`.

**Comment by feross** [comment]
> ts-standard is awesome! We should update the README.md, that everyone using TypeScript should use ts-standard.

Opened an issue for this here: https://github.com/standard/standard/issues/1590

**Comment by toddbluhm** [comment]
`ts-standard` is now officially part of the `standard` github organization.

https://github.com/standard/ts-standard

**Comment by feross** [comment]
Since `ts-standard` is now the official recommendation, I'll close this issue. 

https://github.com/standard/ts-standard
